packages/next/shared/lib/router/router.ts

```diff
@@ -2260,21 +2260,27 @@ interface MiddlewareEffectParams<T extends FetchDataOutput> {
   router: Router
 }
 
-function matchesMiddleware<T extends FetchDataOutput>(
+async function matchesMiddleware<T extends FetchDataOutput>(
   options: MiddlewareEffectParams<T>
 ): Promise<boolean> {
-  return Promise.resolve(options.router.pageLoader.getMiddleware()).then(
-    (middleware) => {
-      const { pathname: asPathname } = parsePath(options.asPath)
-      const cleanedAs = hasBasePath(asPathname)
-        ? removeBasePath(asPathname)
-        : asPathname
-
-      const regex = middleware?.location
-      return (
-        !!regex && new RegExp(regex).test(addLocale(cleanedAs, options.locale))
-      )
-    }
+  const matchers = await Promise.resolve(
+    options.router.pageLoader.getMiddleware()
+  )
```

Comment:
I also don't know that. It's weird but I didn't change since I also was not sure. cc @ijjk

Reply:
In development we have to fetch the middleware manifest since it's generated dynamically, in production the manifest is inlined into the bundle and doesn't need to be fetched.

Reply:
In that case, I wonder why we don't make the `getMiddleware` method asynchronous always